src/tool/xmeta/cpp/symbol_table.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include "models/xmeta_models.h"
+
+namespace xlang::xmeta
+{
+    struct xmeta_symbol_table
+    {
+        xmeta_symbol_table(std::string_view const& idl_assembly_name)
+        {
+            m_assembly_names.emplace_back(idl_assembly_name);
+            m_current_assembly = m_assembly_names.back();
+        }
+
+        // Pushes a namespace to the current namespace scope, and adds it to the symbol table if necessary.
+        void push_namespace(std::string_view const& name, size_t decl_line);
+
+        // Pops a namespace from the namespace scope.
+        void pop_namespace();
+
+        // Clear all values
+        void reset(std::string_view const& assembly_name);
+
+        auto const& get_namespaces() const
+        {
+            return m_namespaces;
+        }
+
+        auto const& get_cur_namespace_body() const
+        {
+            return m_cur_namespace_body;
+        }
+
+        auto const& get_cur_assembly() const
+        {
+            return m_current_assembly;
+        }
+
+        void set_cur_namespace_body(std::shared_ptr<namespace_body_model> const& cur_namespace_body)
+        {
+            this->m_cur_namespace_body = cur_namespace_body;
+        }
+
+        bool semantic_error_exists = false;
+        void write_error(size_t decl_line, std::string_view const& msg);
+        void write_enum_member_name_error(size_t decl_line, std::string_view const& invalid_name, std::string_view const& enum_name);
+        void write_enum_member_expr_ref_error(size_t decl_line, std::string_view const& invalid_name, std::string_view const& enum_name);
+        void write_enum_circular_dependency(size_t decl_line, std::string_view const& invalid_member_id, std::string_view const& enum_name);
+        void write_enum_const_expr_range_error(size_t decl_line, std::string_view const& invalid_expr, std::string_view const& enum_name);
+        void write_namespace_name_error(size_t decl_line, std::string_view const& invalid_name, std::string_view const& original_name);
+        void write_namespace_member_name_error(size_t decl_line, std::string_view const& invalid_name);
+
+    private:
+        std::map<std::string_view, std::shared_ptr<namespace_model>, std::less<>> m_namespaces;
```

Comment:
Wishlist: a method that returns a vector of namesapce_models

Reply:
We can iterate through the map, vector not required